src/main/java/com/puppycrawl/tools/checkstyle/GlobalStatefulModule.java

```diff
@@ -0,0 +1,47 @@
+////////////////////////////////////////////////////////////////////////////////
+// checkstyle: Checks Java source code for adherence to a set of rules.
+// Copyright (C) 2001-2017 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Inherited;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * This annotation means that the module contains global context,
+ * which will be updated across the entire application's lifetime.
+ * across the entire application lifetime. This also means, that all files will be processed
+ * by the same module instance.
+ * This annotation should be used, if a module accumulates some information across the application
+ * lifetime, and processed only once at the end of application lifetime (however, the module still
+ * can produce some messages, while collecting information).
+ * The check methods and fields should be thread safe, because they may be accessed from separate
+ * threads at the same time.
```

Comment:
if state is global, single instance for whole execution, why Check should care about other threads and thread safety at all ?

Reply:
Because it is still called from the separate threads

Reply:
What is a point to create thread(s) for this ?
If there will be two TranslationCheck in config, do you plan to create two threads for them ?

Reply:
> What is a point to create thread(s) for this ?

Threads are created for files, not for checks. Two separate threads, processing two different files might call this check at the same time.

> If there will be two TranslationCheck in config, do you plan to create two threads for them?

If there are two `TranslationCheck` in config, there will be 2 instance of this check, each per configuration. Please refer to my previous answer about threads - "Threads are created for files, not for checks."